src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageFileService.cs

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Specialized;
+using System.Diagnostics;
+using System.IO;
+using System.Net.Http;
+using System.Runtime.Caching;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft;
+using Microsoft.ServiceHub.Framework;
+using Microsoft.ServiceHub.Framework.Services;
+using NuGet.Common;
+using NuGet.PackageManagement.Telemetry;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.VisualStudio.Internal.Contracts;
+using NuGet.VisualStudio.Telemetry;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    public sealed class NuGetPackageFileService : INuGetPackageFileService
+    {
+        public static readonly string IconPrefix = "icon:";
+        public static readonly string LicensePrefix = "license:";
+
+        private ServiceActivationOptions? _options;
+        private IServiceBroker _serviceBroker;
+        private AuthorizationServiceClient? _authorizationServiceClient;
+        private INuGetTelemetryProvider _nuGetTelemetryProvider;
+        private bool _disposedValue;
+
+        internal readonly static MemoryCache IdentityToUriCache = new MemoryCache("PackageSearchMetadata",
+            new NameValueCollection
+            {
+                { "cacheMemoryLimitMegabytes", "4" },
+                { "physicalMemoryLimitPercentage", "0" },
+                { "pollingInterval", "00:02:00" }
+            });
+
+        private static readonly CacheItemPolicy CacheItemPolicy = new CacheItemPolicy
+        {
+            SlidingExpiration = ObjectCache.NoSlidingExpiration,
+            AbsoluteExpiration = ObjectCache.InfiniteAbsoluteExpiration,
+        };
+
+        public static void AddIconToCache(PackageIdentity packageIdentity, Uri iconUri)
+        {
+            string key = NuGetPackageFileService.IconPrefix + packageIdentity.ToString();
+            if (iconUri != null && !IdentityToUriCache.Add(key, iconUri, CacheItemPolicy))
+            {
+                IdentityToUriCache.Remove(key);
+                IdentityToUriCache.Add(key, iconUri, CacheItemPolicy);
+            }
+        }
+
+        public static void AddLicenseToCache(PackageIdentity packageIdentity, Uri embeddedLicenseUri)
+        {
+            Assumes.NotNull(embeddedLicenseUri);
+            string key = NuGetPackageFileService.LicensePrefix + packageIdentity.ToString();
+            if (!IdentityToUriCache.Add(key, embeddedLicenseUri, CacheItemPolicy))
```

Comment:
Why not use Set to avoid the remove/add that happens?

Reply:
fixed in next iteration.

Reply:
(also did for other case in this file.)